Python/optimizer.c

```diff
@@ -590,409 +562,434 @@ translate_bytecode_to_trace(
+    // Failed specialization many times. Deopt!
+    if (tstate->interp->jit_state.specialize_counter >= MAX_SPECIALIZATION_TRIES) {
+        opcode = _PyOpcode_Deopt[opcode];
+    }
+    int rewind_oparg = oparg;
+    while (rewind_oparg > 255) {
+        rewind_oparg >>= 8;
+        target--;
+    }
+
+    int old_stack_level = tstate->interp->jit_state.prev_instr_stacklevel;
+
+    bool needs_guard_ip = _PyOpcode_NeedsGuardIp[opcode];
+    DPRINTF(2, "%p %d: %s(%d) %d %d\n", old_code, target, _PyOpcode_OpName[opcode], oparg, needs_guard_ip, old_stack_level);
+
+#ifdef Py_DEBUG
+    if (oparg > 255) {
+        assert(_Py_GetBaseCodeUnit(old_code, target).op.code == EXTENDED_ARG);
+    }
+#endif
+
+    // Skip over super instructions.
+    if (tstate->interp->jit_state.prev_instr_is_super) {
+        tstate->interp->jit_state.prev_instr_is_super = false;
+        return 1;
+    }
 
-        if (!first && instr == initial_instr) {
-            // We have looped around to the start:
-            RESERVE(1);
+    if (opcode == ENTER_EXECUTOR) {
+        int is_first_instr = tstate->interp->jit_state.close_loop_instr == next_instr || tstate->interp->jit_state.insert_exec_instr == next_instr;
+        if (is_first_instr && tstate->interp->jit_state.code_curr_size > 5) {
             ADD_TO_TRACE(_JUMP_TO_TOP, 0, 0, 0);
             goto done;
         }
+        goto full;
+    }
 
-        DPRINTF(2, "%d: %s(%d)\n", target, _PyOpcode_OpName[opcode], oparg);
+    if (!tstate->interp->jit_state.dependencies_still_valid) {
+        goto done;
+    }
 
-        if (opcode == EXTENDED_ARG) {
-            instr++;
-            opcode = instr->op.code;
-            oparg = (oparg << 8) | instr->op.arg;
-            if (opcode == EXTENDED_ARG) {
-                instr--;
-                goto done;
+    // Strange control-flow, unsupported opcode, etc.
+    if (tstate->interp->jit_state.dynamic_jump_taken) {
+        DPRINTF(2, "Unsupported: dynamic jump taken\n");
+        goto unsupported;
+    }
+
+    // This happens when a recursive call happens that we can't trace. Such as Python -> C -> Python calls
+    // If we haven't guarded the IP, then it's untraceable.
+    if (frame != tstate->interp->jit_state.prev_instr_frame && !needs_guard_ip) {
+        DPRINTF(2, "Unsupported: unguardable jump taken\n");
+        goto unsupported;
+    }
+
+    if (oparg > 0xFFFF) {
+        DPRINTF(2, "Unsupported: oparg too large\n");
+        goto unsupported;
+    }
+
+    // TODO (gh-140277): The constituent use one extra stack slot. So we need to check for headroom.
+    if (opcode == BINARY_OP_SUBSCR_GETITEM && old_stack_level + 1 > old_code->co_stacksize) {
+        goto unsupported;
+    }
+
+    if (opcode == WITH_EXCEPT_START || opcode == RERAISE || opcode == CLEANUP_THROW || opcode == PUSH_EXC_INFO) {
+        DPRINTF(2, "Unsupported: strange control-flow\n");
+        goto unsupported;
+    }
+
+    if (frame->owner >= FRAME_OWNED_BY_INTERPRETER) {
```

Comment:
```suggestion
    assert (frame->owner < FRAME_OWNED_BY_INTERPRETER);
```
This can only happen for the `INTERPRETER_EXIT` instruction, and we should have stopped tracing before calling this function.

Reply:
No it can happen when tracing something that belongs to a C call. For example `sum(# inner python function here)`.

We just need to end up tracing to the function that belongs to the entry frame, which happens often.

Remember that we no longer have the concept of "stack underflow" everything is now fair game.

Reply:
What I'm saying is that `INTERPRETER_EXIT` should stop tracing and not reach here.

Reply:
Yes I realised that, I have a fix that I will push shortly.

Reply:
The problem was that there was no associated code object with the entry frame. I will fix that in this push.